drivers/mipi_dbi/mipi_dbi_rpi_pico_pio.c

```diff
@@ -0,0 +1,507 @@
+ */
+
+#define DT_DRV_COMPAT raspberrypi_pico_mipi_dbi_pio
+
+#include <hardware/pio.h>
+#include <hardware/dma.h>
+
+#include <zephyr/dt-bindings/dma/rpi-pico-dma-common.h>
+#include <zephyr/drivers/mipi_dbi.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(mipi_dbi_pico_pio, CONFIG_MIPI_DBI_LOG_LEVEL);
+
+/* The MIPI DBI spec allows 8, 9, and 16 bits */
+#define MIPI_DBI_MAX_DATA_BUS_WIDTH 16
+
+/* max 4 splits as max 4 state machines per PIO instance */
+#define MIPI_DBI_MAX_SPLITS 4
+
+#define CLK_DIV           CONFIG_MIPI_DBI_RPI_PICO_PIO_CLOCK_DIV
+#define SIDESET_BIT_COUNT 2
+#define CYCLES_PER_BIT    4
+#define SM_SEND_BIT_DURATION_NS                                                                    \
+	((1e9 * CYCLES_PER_BIT * CLK_DIV) / CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)
+
+struct mipi_dbi_pico_pio_config {
+	const struct device *dev_dma;
+	const struct device *dev_pio;
+	struct k_msgq *msq_dma;
+
+	/* Parallel 8080 data GPIOs */
+	const struct gpio_dt_spec data[MIPI_DBI_MAX_DATA_BUS_WIDTH];
+	const uint8_t data_bus_width;
+
+	/* Write (type B) GPIO */
+	const struct gpio_dt_spec wr;
+
+	/* Chip-select GPIO */
+	const struct gpio_dt_spec cs;
+
+	/* Command/Data GPIO */
+	const struct gpio_dt_spec cmd_data;
+
+	/* Reset GPIO */
+	const struct gpio_dt_spec reset;
+};
+
+struct mipi_dbi_pico_pio_dma {
+	int channel;
+	struct dma_config config;
+	struct dma_block_config head_block;
+};
+
+struct mipi_dbi_pico_pio_sm {
+	size_t sm;
+	uint32_t wrap;
+	uint32_t wrap_target;
+	struct pio_program program;
+	uint16_t program_instructions[3];
+};
+
+struct mipi_dbi_pico_pio_split {
+	uint8_t pin_count;
+	int pin_base;
+	int pin_discarded;
+	const struct device *port;
+	struct mipi_dbi_pico_pio_sm sm;
+	struct mipi_dbi_pico_pio_dma dma;
+};
+
+struct mipi_dbi_pico_pio_data {
+	PIO pio;
+	struct k_mutex lock;
+	uint32_t sm_mask;
+	uint8_t split_count;
+	struct mipi_dbi_pico_pio_split split[MIPI_DBI_MAX_SPLITS];
+};
+
+static void mipi_dbi_pio_dma_handler(const struct device *dev, void *user_data, uint32_t channel,
+				     int status)
+{
+	const struct device *dev_mipi_dbi = (const struct device *)user_data;
+	const struct mipi_dbi_pico_pio_config *config = dev_mipi_dbi->config;
+	struct mipi_dbi_pico_pio_data *data = dev_mipi_dbi->data;
+
+	for (int i = 0; i < data->split_count; ++i) {
+		if (data->split[i].dma.channel == channel) {
+			k_msgq_put(config->msq_dma, &channel, K_NO_WAIT);
+		}
+	}
+}
+
+static int mipi_dbi_pio_setup_dma(const struct device *dev, struct mipi_dbi_pico_pio_split *split)
+{
+	split->dma.channel = dma_claim_unused_channel(true);
+	if (split->dma.channel == -1) {
+		return -EPROTO;
+	}
+
+	struct mipi_dbi_pico_pio_data *data = dev->data;
+
+	switch (split->sm.sm) {
+	case 0:
+		split->dma.config.dma_slot =
+			data->pio == pio0 ? RPI_PICO_DMA_SLOT_PIO0_TX0 : RPI_PICO_DMA_SLOT_PIO1_TX0;
```

Comment:
In the case of rp2350, there are three PIOs, so please consider supporting it if possible.